argoproj-labs/argocd-operator
# feat: Add conversion webhook for ArgoCD v1alpha1 to v1beta1 migration

**What type of PR is this?**
/kind enhancement

**What does this PR do / why we need it**:
This PR deprecates ArgoCD `v1alpha` CRD and adds a new ArgoCD `v1beta1` CRD along with conversion webhook to automate migration of ArgoCD v1alpha1 resources to v1beta1.

Changes:
- Add new ArgoCD v1beta1 crd.
- Mark ArgoCD v1alpha1 as deprecated & add back the removed dex & sso deprecated fields in ArgoCD v1alpha1.
- Add explicit conversion funcs for converting  ArgoCD v1alpha1 to v1beta1.
- Configure conversion webhook server to run on 9443 port in operator-controller-manager deployment.
- Disable single namespace install for operator. OLM doesn't support single namespace installation mode for operators with conversion webhooks.
- Use server side validation for "kubectl apply" as client side apply results into failure due to exceeding annotation size limit.
- Add kuttl e2e tests
- Add documentation for manual install of operator with conversion webhook support.

Follow-up PRs (yet to be created):
- Add CI job for testing conversion webhook
- Replace all ArgoCD v1alpha1 references with v1beta1 in codebase
- Update api docs

**Have you updated the necessary documentation?**

* [x] Documentation has been updated. (**Partially**)

**How to test changes / Special notes to the reviewer**:
Operator Installation:
- Follow `tests/olm/README.md` from this PR to install the operator via OLM. 

Manual Test:
- Create a ArgoCD with v1alpha1 version 
    ```bash
    kubectl apply -f - <<EOF
    apiVersion: argoproj.io/v1alpha1
    kind: ArgoCD
    metadata:
      name: argocd
    spec:
      dex:
        config: test-config
    EOF
    ```
- Observe deprecation warning for `kubectl apply`
    `Warning: ArgoCD v1alpha1 is deprecated, please use v1beta1 instead.`

- Verify that conversion to v1beta1 didn't result in data loss
    ```bash
    $ kubectl get argocd/argocd -o yaml
    apiVersion: argoproj.io/v1beta1
    kind: ArgoCD
    metadata:
      annotations:
        kubectl.kubernetes.io/last-applied-configuration: |
          {"apiVersion":"argoproj.io/v1alpha1","kind":"ArgoCD","metadata":{"annotations":{},"name":"argocd","namespace":"default"},"spec":{"dex":{"config":"test-config"}}}
      name: argocd
    spec:
      sso:
        dex:
          config: test-config
        provider: dex
    ```

Automated Test:
- `kubectl kuttl test ./tests/olm --config ./tests/kuttl-tests.yaml`
